File: day53/generateParenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []
        self._gen(n, n, "")
        return self.res

    def _gen(self, left, right, res):
            if right == 0:
                self.res.append(res)
                return
                
            if left != 0: self._gen(left - 1, right, res + "(")
            if right > left: self._gen(left, right - 1, res + ")")

File: day53/generateParenthesis.java
/*
22. 括号生成
 n 代表生成括号的对数，请你设计一个函数，用于能够生成所有可能的并且 有效的 括号组合。

https://leetcode-cn.com/problems/generate-parentheses/
*/

class Solution {
    public List<String> res;

    public List<String> generateParenthesis(int n) {
        this.res = new LinkedList<>();
        genOne(n, n, "");
        return this.res;
    }

    private void genOne(int left, int right, String subList) {
        if (right == 0) {
            this.res.add(subList);
            return;
        }

        if (left != 0) genOne(left - 1, right, subList + "(");
        if (right > left) genOne(left, right - 1, subList + ")");
    }
}
